c/cert/test/rules/ERR32-C/test.c

```diff
@@ -0,0 +1,42 @@
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+
+typedef void (*pfv)(int);
+
+void handler1(int signum) {
+  pfv old_handler = signal(signum, SIG_DFL);
+  if (old_handler == SIG_ERR) {
+    perror(""); // NON_COMPLIANT
+  }
+}
+
+void handler2(int signum) {
+  pfv old_handler = signal(signum, SIG_DFL);
+  if (old_handler == SIG_ERR) {
+    abort(); // COMPLIANT
+  }
+}
+
+void handler3(int signum) { pfv old_handler = signal(signum, SIG_DFL); }
+
+pfv helper4(int signum) {
+  perror(""); // NON_COMPLIANT
+  return signal(signum, SIG_DFL);
+}
+void handler4(int signum) { pfv old_handler = helper4(signum); }
+
+int main(void) {
```

Comment:
is this also NON_COMPLIANT?
```
int main(void) {
  pfv old_handler = signal(SIGINT, handler1);
  if (old_handler == SIG_ERR) {
    perror(""); // NON_COMPLIANT
  }

  old_handler = signal(SIGINT, handler2);

  old_handler = signal(SIGINT, handler3);

  old_handler = signal(SIGINT, handler4);

  old_handler = signal(SIGINT, handlerX);

  FILE *fp = fopen("something", "r");
   if( fp == NULL ) {
      perror("Error: "); // NON_COMPLIANT?
   }
}
```
either way this one generates alerts for results for every matching call to signal above, do we want to consider just the closest above? (although with no results on openpilot this may actually not be an alert explosion issue at all, your call!)

Reply:
er, sorry just to clarify I mean this one:
```
FILE *fp = fopen("something", "r");
   if( fp == NULL ) {
      perror("Error: "); // NON_COMPLIANT?
   }
```
fopen will "re-determinate" the errno? or is it possible that the errno will still have been last used for something related to a signal event... if the signal was seen after the fopen call ...